msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md

```diff
@@ -206,33 +206,44 @@ If you need to access Microsoft Graph data, configure your server-side code to:
-3. When the exchange fails, you must ask for consent. Use the user interface (UI) to ask the app user to grant other consent. This UI must include a button that triggers an Azure AD consent dialog using [Silent authentication](~/concepts/authentication/auth-silent-aad.md).
-4. When asking for more consent from Azure AD, you must include `prompt=consent` in your [query-string-parameter](~/tabs/how-to/authentication/auth-silent-aad.md#get-the-user-context) to Azure AD, otherwise Azure AD wouldn't ask for other scopes.
-    - Instead of `?scope={scopes}`, use `?prompt=consent&scope={scopes}`
-    - Ensure that `{scopes}` includes all the scopes you're prompting the user for, for example, `Mail.Read` or `User.Read`.
+    1. The token retrieved using `getAuthToken()` must be exchanged on the server-side using Azure AD [on-behalf-of flow](/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow) to get access to those other Graph APIs. Ensure you use the v2 Graph endpoint for this exchange.
+    2. If the exchange fails, Azure AD returns an invalid grant exception. It usually responds with one of the two error messages, `invalid_grant` or `interaction_required`.
+    3. When the exchange fails, you must ask for consent. Use the user interface (UI) to ask the app user to grant other consent. This UI must include a button that triggers an Azure AD consent dialog using [Silent authentication](~/concepts/authentication/auth-silent-aad.md).
+    4. When asking for more consent from Azure AD, you must include `prompt=consent` in your [query-string-parameter](~/tabs/how-to/authentication/auth-silent-aad.md#get-the-user-context) to Azure AD, otherwise Azure AD wouldn't ask for other scopes.
+        - Instead of `?scope={scopes}`, use `?prompt=consent&scope={scopes}`
+        - Ensure that `{scopes}` includes all the scopes you're prompting the user for, for example, `Mail.Read` or `User.Read`.
 
-    To handle incremental consent for tab app, see [incremental and dynamic user consent](/azure/active-directory/develop/v2-permissions-and-consent).
-5. After the app user has granted more permissions, retry the OBO flow to get access to these other APIs.
+        To handle incremental consent for tab app, see [incremental and dynamic user consent](/azure/active-directory/develop/v2-permissions-and-consent).
+    5. After the app user has granted more permissions, retry the OBO flow to get access to these other APIs.
     </details>
 
+1. Delay in fetching access token for Graph permission and scope: If your app uses staggered permission for Graph permissions and scopes, the app user must give consent again to fetch the access token. In this scenario, when the app user gives consent, a middle-tier service runs the OBO flow. For more information, see the Step 2 of [exchange the token ID with the server-side token](tab-sso-graph-api.md#exchange-the-token-id-with-the-server-side-token).
```

Comment:
I think this, reads weirdly. 

I took the liberty to rewrite it a bit to make it clearer, feel free to change where necessary but I hope this makes sense

> Race condition in fetching Graph access token via on-behalf-of (OBO) flow after consent: If your app calls Microsoft Graph, you might use the [on-behalf-of (OBO) flow in you API](tab-sso-graph-api.md#exchange-the-token-id-with-the-server-side-token) to get a valid Graph token for that user.
> 
> If a user has not granted your AAD application consent for these scopes before your OBO call will fail with an `invalid_grant` or `interaction_required` error, telling you that you need to prompt the user for their consent. 
> 
> When that user has provided their consent and you try to make an OBO call immediately, sometimes there is a race condition between Azure AD propagating this consent and the OBO request taking place. This can lead to your OBO call failing with the same `invalid_grant` or `interaction_required` errors.
> 
> If your application is unaware of this behaviour, it might ask the user for consent multiple times.
> 
> There's no workaround to this limitation, Azure AD recommends that you can build a meaningful wait-and-retry mechanism to overcome this issue.
> 
> This wait-and-retry mechanism should keep track if a user has consented to the required scopes. Then if an API call that would include an OBO request fails with the above errors, but the user has already consented, avoid showing the consent prompt to the user. Instead, wait a brief period before retrying the API call. Usually, Azure AD will have propagated the consent with three to five seconds. [In one of our sample applications](https://github.com/OfficeDev/Microsoft-Teams-Samples/blob/8f266c33608d6d7b4cf89c81779ccf49e7664c1e/samples/bot-tab-conversations/csharp/Source/ConversationalTabs.Web/ClientApp/src/utils/UtilsFunctions.ts#LL8C1-L8C1), we retry up to three times with the wait doubling between each retry, starting at a one second wait.
> 
> If after three to five retries the OBO flow is still failing, the user might not have consented to all the required scopes, and you may have to prompt them to consent again.
> 
> This approach helps reduce the likelihood that the user is not asked to consent to the same details more than once.

Reply:
@eoinobrien Updated the doc as suggested

Reply:
Updated the doc as suggested with minor updates.